### DISCUSSION TITLE
How can I hide the Dock icon on MacOS? `Info.plist` not working

### DISCUSSION THREAD

**Discussion by Irasil** [AvaloniaUI/Avalonia#13568] [Q&A] [answered]
I have read that I need to set `LSUIElement` to `YES` in the `Info.plist` to prevent the icon from appearing on the Dock [Documentation Archive](https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/LaunchServicesKeys.html#//apple_ref/doc/uid/20001431-108256).

After testing this, I noticed that the Info.plist has no effect on the app. No matter what I set CFBundleName and/or CFBundleDisplayName to, the app retains the name I assigned in app.axaml. This should not be the case according to this discussion (https://github.com/AvaloniaUI/Avalonia/discussions/9346). What am I doing wrong?

Here is the structure of the app:

![Screenshot 2023-11-10 122119](https://github.com/AvaloniaUI/Avalonia/assets/91846030/2e46c20f-a6f8-40f8-8e30-f983a42ec86b)

The .Avalonia project is for the UI; here, I have also assigned the name in the app.axaml file, and .Avalonia.MacOS for all platform-specific aspects. I created the Info.plist in the MacOS project.
For creating the .Avalonia.MacOS project, I used the "AvaloniaUI .Net App" Template and changed the TFM to `net7.0-macos`, referencing the UI project.

contents of the Info.plist:

```xml
<dict>
	<key>CFBundleDisplayName</key>
	<string>Hallo</string>
	<key>CFBundleExecutable</key>
	<string>IBI.aws.Avalonia.MacOS</string>
	<key>CFBundleIdentifier</key>
	<string>com.myapp.macos</string>
	<key>CFBundleName</key>
	<string>MyApp</string>
	<key>LSUIElement</key>
	<true/>
	<key>MonoBundleExecutable</key>
	<string>IBI.aws.Avalonia.MacOS.exe</string>
</dict>
```

Even when I publish the app and customize the `Info.plist` in the .app, the name of the app remains unchanged. Does the `Info.plist` have no impact on Avalonia apps?

Or is there another way to hide the Dock icon without using `Info.plist`?

**Comment by maxkatz6** [comment]
It has, if Info.plist is assigned right, it should work with Avalonia app just like with any other .NET app, there is zero difference.

The only non-standard thing that Avalonia does with the app title - is setting app title in runtime via [changing process name](https://github.com/AvaloniaUI/Avalonia/blob/master/native/Avalonia.Native/src/OSX/main.mm#L12). It might change dock tile name as well, but not in the applications list I suppose. Also should affect the icon.

I can't help what's exactly wrong with LSUIElement though.

**Reply by thevortexcloud** [reply]
> I can't help what's exactly wrong with LSUIElement though.

We worked it out. Avalonia has a bunch of code which seems to override the option at runtime (Avalonia has its own platform specific option to manage it). If that should be considered a bug or not I am not sure.

**Comment by Irasil** [comment]
Here's what I found out so far:

I created a new Avalonia Desktop project on Windows. I cloned it on Mac and executed `dotnet publish -r osx-x64 --configuration Release -p:UseAppHost=true`. After that, I copied the `Info.plist` from the [documentation](https://docs.avaloniaui.net/docs/next/deployment/macOS#manual) and changed `CFBundleExecutable` to my executable. I used the shell script to create a .app. When I run the app, the title is "Avalonia Application."

First change in the project: 
I added `Name="Test"` in app.axaml. After publishing and running the script, the title is "MyApp," as defined in `Info.plist`. That's not bad :) However, if I add `LSUIElement`, it is ignored.

For my application, I need a project with the TFM `net7.0-macos` to access native APIs. If I change the TFM of the test project described above to `net7.0-macos`, I also have to change `<OutputType>` from `WinExe` to `Exe`. When I publish, I only get a `.pkg` file. Therefore, I can't apply the script to create an app bundle. If I look in `/bin/Release/net7.0-macos/`, I see that a .app bundle was created. It can be launched but has the title "Avalonia Application" again. When I open the .app bundle via "Show package content," I see that an `Info.plist` was automatically created, with the project name defined as the Name. So, as soon as I set the TFM to net7.0-macos, the `Info.plist` is ignored.

However, if I create a .Net Mac App project on Mac, which also has `net7.0-macos` as the TFM, and publish it with the same command, it doesn't create a .pkg, it create an .app, which doesn't ignore the `Info.plist`, and `LSUIElement` works right away.

Can someone explain why Avalonia + `net7.0-macos` results in a `.pkg` file, and is there a way to bypass this?

I'm new to programming and don't have much experience, so I would appreciate any hints on what I might be doing wrong. I'm really desperate...

Thanks in advance.

**Reply by thevortexcloud** [reply]
It's hard to say without seeing your project file.

EDIT: Does this issue help?

https://github.com/dotnet/sdk/issues/27841

**Reply by Irasil** [reply]
Nice, that helped. So, I was able to create a .app again using the script. When I launch the created .app, the name I defined in the Info.plist appears. Thank you!

Unfortunately, `LSUIElement` is still being ignored. I'm now looking for other possibilities, but I appreciate any tips.

[Here ](https://github.com/Irasil/AvaloniaPListTest) is the link to my test project.

**Reply by thevortexcloud** [reply]
I am seeing conflicting information about `LSUIElement` in some examples it's a boolean value. In others it's a string where you set the value to `1`. In Apple's own documentation it's a boolean that can be set to a `YES` value (what ever that is, I am guessing a string), and some other ones it's a a string value of `true`. 

> LSUIElement (Boolean - macOS) specifies whether the app runs as an agent app. If this key is set to YES, Launch Services runs the app as an agent app. Agent apps do not appear in the Dock or in the Force Quit window. Although they typically run as background apps, they can come to the foreground to present a user interface if desired. A click on a window belonging to an agent app brings that app forward to handle events.

I am not clear why this is so inconsistent but one of them probably works.

**Reply by Irasil** [reply]
I hate to say this, but I've also come across this inconsistency and have already tried both options without success.

I have tested these options:

```
<key>LSUIElement</key>
<string>YES<string/>
<string>yes<string/>
<string>true<string/>
<string>1<string/>
```

With Boolean, I can only test `<true/>` since the `<boolean>` tag is not recognized.

Furthermore, I found out that `LSRequiresIPhoneOS` is also ignored.

**Reply by thevortexcloud** [reply]
Are you trying those all at once? It's a key value list from what I can tell. So each key is followed by a single corresponding value.

Also apparently you are not the first person to have this problem?

https://github.com/AvaloniaUI/Avalonia/issues/4849

It sounds like you can actually set this value at runtime using a C API. Which may work better. I am not sure. In any case you would need to use `DllImport` for this.

That said if you are just trying to hide the dock icon, there is apparently a Mac specific Avalonia launch option for that. Which might actually be the problem if it's overriding this for you (the default value is to show the dock icon).

https://github.com/AvaloniaUI/Avalonia/blob/0c4f372410d0f45c814930da84c73f65b377176b/src/Avalonia.Native/AvaloniaNativePlatformExtensions.cs#L87C5-L91C1

From reading [the Avalonia native code ](https://github.com/AvaloniaUI/Avalonia/blob/0c4f372410d0f45c814930da84c73f65b377176b/native/Avalonia.Native/src/OSX/main.mm#L132C1-L142C6)it looks like it just sets this value https://developer.apple.com/documentation/appkit/nsapplicationactivationpolicy/nsapplicationactivationpolicyaccessory

Which apparently is the same as `LSUIElement`. So I am guessing this will solve your problem.

**Reply by Irasil** [reply]
No, I tested the different values one after the other :)

I appreciate this information, and I think that will be the solution.

Unfortunately, this is all new to me, and I don't quite understand what I need to do. Am I correct in understanding that I simply need to set [ShowInDock](https://github.com/AvaloniaUI/Avalonia/blob/0c4f372410d0f45c814930da84c73f65b377176b/src/Avalonia.Native/AvaloniaNativePlatformExtensions.cs#L87C5-L91C1) to false?

If so, I don't know how to do that. I searched for `Avalonia.Native.OSX.xcodeproj`, but I can't find it. Do I need to download the Avalonia repository and then set ShowInDock to false there? And then add my edited dylib in the AppBuilder afterward? As described here [macOS Development](https://docs.avaloniaui.net/docs/next/guides/platforms/macos-development)?

**Reply by thevortexcloud** [reply, marked as answer]
> Unfortunately, this is all new to me, and I don't quite understand what I need to do. Am I correct in understanding that I simply need to set [ShowInDock](https://github.com/AvaloniaUI/Avalonia/blob/0c4f372410d0f45c814930da84c73f65b377176b/src/Avalonia.Native/AvaloniaNativePlatformExtensions.cs#L87C5-L91C1) to false?

If I am reading all the code correctly, yes. I however still have no Mac to test on or any real experience programming for Macs.

> If so, I don't know how to do that. 

Avalonia allows for setting a bunch of platform specific options during Launch. Typically you do something like 

```csharp
class Program {
    // Initialization code. Don't use any Avalonia, third-party APIs or any
    // SynchronizationContext-reliant code before AppMain is called: things aren't initialized
    // yet and stuff might break.
    [STAThread]
    public static void Main(string[] args) => BuildAvaloniaApp()
        .StartWithClassicDesktopLifetime(args);

    // Avalonia configuration, don't remove; also used by visual designer.
    public static AppBuilder BuildAvaloniaApp()
        => AppBuilder.Configure<App>()
            .UsePlatformDetect()
            .WithInterFont()
            .LogToTrace()
            .UseReactiveUI()
            .With(new MacOSPlatformOptions() { ShowInDock = false});
}
```

**Reply by Irasil** [reply]
If you don't have a cape yet, you definitely need one. You're a fucking hero!
I thank you endlessly!

**Reply by thevortexcloud** [reply]
Haha, your welcome! This however is something that really should be documented on the Avalonia docs site (along with all the other platform options). It's not even mentioned you can set platform options anywhere that I can see.

**Reply by timunie** [reply]
any contribution to the docs are welcome

**Reply by Linlccc** [reply]
How should I dynamically configure whether to display during program execution?

**Reply by thevortexcloud** [reply]
Right now I don't think you can reconfigure the platform options once Avalonia has started. One of the Avalonia devs can probably say otherwise though. You can however likely just do exactly what Avalonia is doing and use the native API directly. I posted the links to most of it earlier in the thread.
